processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -157,18 +184,35 @@ func (f internalFilterRegexp) labelMatched(metric *metricspb.Metric) bool {
 			}
 
 			keyFound = true
-			if len(metric.Timeseries) > 0 && !value.MatchString(metric.Timeseries[0].LabelValues[idx].Value) {
-				return false
-			}
+			labelIndexValueMap[idx] = value
 		}
 
-		// if a label-key is not found then return false only if the given label-value is non-empty. If a given label-value is empty
-		// and the key is not found then return true. In this approach we can make sure certain key is not present which is a valid use case.
+		// if a label-key is not found then return nil only if the given label-value is non-empty. If a given label-value is empty
+		// and the key is not found then move forward. In this approach we can make sure certain key is not present which is a valid use case.
 		if !keyFound && !value.MatchString("") {
-			return false
+			return nil
+		}
+	}
+
+	for _, timeseries := range metric.Timeseries {
+		allValuesMatched := true
```

Comment:
Looks like there's a lot of duplicated code, can you clean it up?

Reply:
Agreed.  I think much of this commonality could be extracted by defining a `StringMatcher` interface:

```go
type StringMatcher interface {
	MatchString(string) bool
}
```

All of the values in `internalFilterRegexp` will already satisfy it since `Regexp` has a `MatchString(string)` method.  `internalFilterStrict` can be made to utilize it by defining a value type:

```
type strictMatcher string
func (s strictMatcher) MatchString(cmp string) bool {
	return string(s) == cmp
)
```

At that point the `labelMatched()` logic can be extracted and updated to this signature:

```go
labelMatched(labelMatchers []StringMatcher, metric *metricspb.Metric) *metricspb.Metric {...}
```

Since the only value used from the receiver in either of the `labelMatched` implementations is the list of match targets, this should neatly eliminate the duplication.

Reply:
Tried to follow your suggestion. Pushed an update.

Reply:
Copied your suggested code @Aneurysm9 . Please check.